docs/source/extending_pywr/extending_pywr_nodes.rst

```diff
@@ -0,0 +1,116 @@
+
+Extending Pywr with custom Nodes
+--------------------------------
+
+Nodes and subclasses thereof provide the basic network structure in Pywr. There are several different types
+of node available. Two major categories exist: flow nodes and storage nodes. Flow nodes are the typical nodes
+that represent rivers, pipes and other features from, through or to which a resource can flow. These nodes are
+typically characterised by minimum and maximum flow rates. Storage nodes provide the ability to store resource
+from one time-step to another, and are characterised by minimum, maximum and initial volumes.
+
+There are three fundamental types of node in Pywr:
+
+* ``Input`` nodes add water to the system
+* ``Output`` nodes remove water from the system
+* ``Link`` nodes do not add or remove water from the system
+
+There is a fourth node type, ``Storage``, which can be considered fundamental because there are special rules for
+it's behaviour in the linear programme used to solve the water balance:
+
+* ``Storage`` nodes can carry water from one timestep to the next
+
+All other node types in Pywr are subclasses of these base types. For example, the ``pywr.nodes.Catchment`` node type
+is a special case of ``Input`` where the ``min_flow`` and ``max_flow`` properties are equal.
+
+The most common way to create a new node type is using a compound node. A compound node contains one or more existing
+nodes and is used to manage common or more complex arrangements of the basic node types. An example of a compound node
+is the ``PiecewiseLink``. It is composed of a link (``OUT_1``) which receives water from upstream and an link
+(``IN_1``) which conveys water downstream, connected by a set of links in parallel (``LNK_1`` ... ``LNK_N``) each with
+a different ``max_flow`` and ``cost``, illustrated below::
+
+                               /-->-- LNK_1 -->--\
+    UPSTREAM -->-- OUT_1 -->--|--->--  ...  -->---|-->-- IN_1 -->-- DOWNSTREAM
+                               \-->-- LNK_N -->--/
+
+Let's look at an example to create a new node type that represents a leaky pipe. To remove water from the system we
+need to use an output node (``LEAK``), with two links representing the boundaries of the compound node (``INFLOW`` and
+``OUTFLOW``)::
+
+    UPSTREAM -->-- INFLOW -->-- OUTFLOW -->-- DOWNSTREAM
+                     |
+                     \------>--  LEAK
+
+This is a simple structure which represents leakage as a demand with a maximum value and a benefit to be supplied. It
+is slightly flawed as the leakage volume does not vary proportionally to flow through the link, but is sufficient as
+an example:
+
+.. code-block:: python
+
+    from pywr.nodes import Node, Link, Output
+
+    class LeakyPipe(Node):
+        def __init__(self, leakage, leakage_cost=-99999, *args, **kwargs):
+            self.allow_isolated = True  # Required for compound nodes
+
+            super(LeakyPipe, self).__init__(*args, **kwargs)
+
+            # Define the internal nodes
```

Comment:
Maybe comment about the importance of the `parent=self` keyword.

Reply:
I don't actually remember what this is myself...

Reply:
I think it may only be used in the `draw_graph` thing, but it's a good thing to know about a particular node I guess.